stargz/keychain/kubeconfig.go

```diff
@@ -0,0 +1,270 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package keychain
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	dcfile "github.com/docker/cli/cli/config/configfile"
+	dctypes "github.com/docker/cli/cli/config/types"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const dockerconfigSelector = "type=" + string(corev1.SecretTypeDockerConfigJson)
+
+type KubeconfigKeychainConfig struct {
+	KubeconfigPath string `toml:"kubeconfig_path"`
+}
+
+// NewKubeconfigKeychain provides a keychain which can sync its contents with
+// kubernetes API server by fetching all `kubernetes.io/dockerconfigjson`
+// secrets in the cluster with provided kubeconfig. If specified kubeconfig path
+// is "", this keychain disables secrets synchronization. It's OK that config
+// provides kubeconfig path but the file doesn't exist at that moment. In this
+// case, this keychain keeps on trying to read the specified path periodically
+// and when the file is actually provided, this keychain tries to access API
+// server using the file. This enables us to lazily provide kubeconfig file to
+// this snapshotter.
+// TODO: support update of kubeconfig file
+func NewKubeconfigKeychain(ctx context.Context, config KubeconfigKeychainConfig) authn.Keychain {
+	var (
+		kubeconfigPath = config.KubeconfigPath
+	)
+	if kubeconfigPath == "" { // If no path provided, disable syncing.
+		log.G(ctx).Warn("No kubeconfig provided; Disabling sync")
```

Comment:
No support for `$KUBECONFIG`?

Reply:
Thanks for the comment. I added `KUBECONFIG` support.